### PR TITLE
Stop propagation on `withConstrainedTabbing`

### LINKED ISSUE #26115: Broken keyboard navigation within text color toolbar
**Describe the bug**
Cannot navigate to hue selector or text input field within the custom color selector from the toolbar (see screenshot 2).

**To reproduce**
1. Type in some paragraph text
2. Open up custom color selector from toolbar (can be done with the keyboard)
3. Arrow keys correctly move the brightness/saturation selector
4. Tab does not navigate to the hue selector

**Expected behavior**
Pressing tab within the custom color selector should navigate to the hue selector (and then another tab should navigate to the text input, and so on).

**Screenshots**
text color toolbar menu
text color toolbar's custom color popover

**Editor version:**
- WordPress version: 5.5.1
- Does the website has Gutenberg plugin installed, or is it using the block editor that comes by default? Gutenberg plugin
- If the Gutenberg plugin is installed, which version is it? Master branch (b126b3f)

**Desktop:**
 - OS: MacOS
 - Browser: Chrome
 - Version: 85.0.4183.121

### PR DESCRIPTION
**Pull request by ajlende**
## Description

Fixes #26115

Nested `withConstrainedTabbing` components were causing focus to more to the outermost parent component, for example, in the custom text color options from the paragraph toolbar.

## How has this been tested?

Try tabbing through nested popovers such as Paragraph > More rich text controls > Text color > Custom color.

## Screenshots 

Before:
color-picker-before

After:
color-picker-after

## Types of changes

Bug Fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/higher-order/with-constrained-tabbing/index.js [hunk 1]
```diff
@@ -30,6 +30,12 @@ const withConstrainedTabbing = createHigherOrderComponent(
 				const firstTabbable = tabbables[ 0 ];
 				const lastTabbable = tabbables[ tabbables.length - 1 ];
 
+				if ( tabbables.includes( event.target ) ) {
+					// Stops nested withConstrainedTabbing components from always focusing
+					// the outermost component.
+					event.stopPropagation();
+				}
```
Comment (youknowriad) [inline packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: Hi there!

i've been thinking a bit here, it seems we're relying on React event handlers for this HoC. the issue with that is that it has unexpected behaviors when using "Portals":

 1- It will catch events from portals even if these portals are not in the same DOM tree.
 2- it won't catch events from portals that are on the same DOM tree but not on the same React tree. (could happen with slots for instance)

I think the current solution solves 1 but doesn't solve 2.

So in reality for this HoC, we only care about the DOM tree and not the React tree. So I believe a better fix might be to change the event handler to use a `ref` to attach a DOM event directly and not rely on the `onKeyDown` prop.

Does that make sense?

Comment (ajlende) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: Hey, thanks for taking a look! 🙂 

What you said makes sense. However, the problem I'm trying to solve isn't related to portals, it just happens to be that `withConstrainedTabbing` is used most often with portals. Take this simple example:

```jsx
const InnerTabbing = withConstrainedTabbing( () => {
	return (
		<div>
			<Button>1</Button>
			<Button>2</Button>
			<Button>3</Button>
		</div>
	);
} );

const TabbingExample = withConstrainedTabbing( () => {
	const [ showInner, setShowInner ] = useState( false );
	return (
		<div>
			<Button onClick={ () => setShowInner( ( prev ) => ! prev ) }>
				Toggle
			</Button>
			{ showInner && <InnerTabbing /> }
		</div>
	);
} );
```

tabbing-example

I would expect the constrained tabbing to wrap around to the numbered buttons, not the toggle button. And that's what this PR solves.

Comment (youknowriad) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: Well I'd personally say, that this example should probably never exist because even if tabbing is wrapped inside the buttons, you'll create a bad interaction:

- you can tab from outside to inside "InnerTabbing"  and then you get stuck there. I believe that's an a11y problem.

So IMO, both ways are broken interactions in this case, in fact, I believe it's better to go out in this case and just ignore the "inner constrained tabbing" to avoid "tab traps".

Comment (ajlende) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: That's a fair point. I may have oversimplified my example a little too much—I wanted to show just the `withConstrainedTabbing` part in isolation and didn't include any ways to exit the tabbing. Most places in the codebase use `withConstrainedTabbing` alongside `withFocusReturn` and some way to unmount the component so that focus can return to wherever it was before. My example should have done the same.

If you're adding this component in a real-world scenario, you'd probably quickly realize that you have to add a way to exit as well since you're already modifying how tabbing is handled. But, just as `withConstrainedTabbing` can be used without a portal, exiting the loop can be done without `withFocusReturn` even if that is the most common way to do it.

My worry is that someone might unknowingly nest a component that uses `withConstrainedTabbing` inside another component that uses `withConstrainedTabbing` like was done with the color picker (#26115) but outside of the portal context. Changing the code to use DOM events instead of React's events may fix the specific case for portals, but it doesn't solve it generally everywhere.

Furthermore, I'd personally expect a component like this to use React events over DOM events anyway—matching the structure of the components in code rather than figuring out where in the DOM the code will render and tracing the events that way. Although, people's expectations on that will probably go both ways.

Comment (youknowriad) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: > Furthermore, I'd personally expect a component like this to use React events over DOM events anyway

Why do you think so? For me, it's the exact opposite because this component is specifically about focus handling which is very DOM related and taking React tree into consideration here doesn't really make sense to me.

 > If you're adding this component in a real-world scenario, you'd probably quickly realize that you have to add a way to exit as well since you're already modifying how tabbing is handled. 

I think adding a way to exit (say escape) is not a good solution here, because you get into the tab trap using "tab" and not by opening something or doing a more explicit action which for me indicates that it's a tab trap https://www.w3.org/TR/UNDERSTANDING-WCAG20/keyboard-operation-trapping.html and shouldn't be possible at all. A developer adding withConstrainedTabbing nested will quickly discover that it doesn't work (which is good because what he did is wrong). 

A solution might be to actually trigger a warning or an error if we're able to detect it.

Comment (ajlende) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: > Furthermore, I'd personally expect a component like this to use React events over DOM events anyway
> > Why do you think so? For me, it's the exact opposite because this component is specifically about focus handling which is very DOM related and taking React tree into consideration here doesn't really make sense to me.

In React, JSX obscures the fact that you're actually just writing JavaScript, not HTML. But because they're so similar, they're expected to behave the same. Unfortunately, there's a contradiction between the JSX and HTML when you're using portals.

When I'm writing code in React, I don't want to have to worry if a component is rendered inside a portal or not—things should work the same either way. The preferred method for handling events in React is with the React event system (you don't use a `ref` for every event you want to attach), so it's breaking away from the normal case by using a direct `ref` to the underlying DOM.

So, I guess the core reason that the React system is more natural to me is because we're writing React code, not HTML.

Comment (youknowriad) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: That's reasonable but the problem is that `withConstrainedTabbing`  is specifically feature intended for DOM and HTML and not something as abstract as React. For instance, it has no meaning in the mobile environment.

Basically I found that for all the hooks/code that deals with focus management and web a11y, it's always better to rely on DOM events. 

In fact, I saw some folks raise this exact same issue to Dan Abramov on twitter about whether the virtual event bubbling used by React when you use portals is a good or bad thing. I'd be surprised if they change that in a future breaking change.

Comment (ajlende) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: Still, the official documentation explicitly talks about event bubbling in portals (https://reactjs.org/docs/portals.html#event-bubbling-through-portals)

> Even though a portal can be anywhere in the DOM tree, it behaves like a normal React child in every other way. Features like context work exactly the same regardless of whether the child is a portal, as the portal still exists in the _React tree_ regardless of position in the _DOM tree_.
>
> This includes event bubbling. An event fired from inside a portal will propagate to ancestors in the containing _React tree_, even if those elements are not ancestors in the _DOM tree_.

Won't it be confusing for folks who land on the official documentation that this behaves differently?

Comment (mcsf) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: > Won't it be confusing for folks who land on the official documentation that this behaves differently?

We can easily clarify that at the very top of our affected focus-related utilities.

Comment (ajlende) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: > > Won't it be confusing for folks who land on the official documentation that this behaves differently?
>
> We can easily clarify that at the very top of our affected focus-related utilities.

Yes, and we'd _have_ to make that clarification in the docs. My point was that using React events is more natural/expected than using DOM events, even with portals.

> Basically I found that for all the hooks/code that deals with focus management and web a11y, it's always better to rely on DOM events.

I still don't quite follow why using DOM events is any better than React events for focus management and web a11y. @youknowriad are there any specific examples you can show me to help me better understand? I appreciate your patience—it just doesn't seem right for me to do something without understanding it first.

If we do go down the DOM event path, it seems that this HoC should really be built-in to the portal (maybe as a boolean `constrainTabbing` prop or something) so it can't be used in a context that isn't supported ([like my earlier example](#discussion_r523207764)). The DOM events option only works in contexts where you're rendering outside of the tree that the component lives. I'd much prefer that to trying to detect misuse and printing a warning.

Comment (youknowriad) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: > I still don't quite follow why using DOM events is any better than React events for focus management and web a11y. @youknowriad are there any specific examples you can show me to help me better understand? I appreciate your patience—it just doesn't seem right for me to do something without understanding it first.

Say you're inside a "menu" and you have buttons rendered in that menu. some of these buttons might be rendered in the same React tree, but you also have a "Slot" on that menu to allow plugins to add their buttons there. You  want the tabbing to work properly inside that menu, to go from one button to another properly and not skip the ones that come from the slot. If you use React events, you won't catch the events that come from these buttons because they'll bubble to the React tree where they are actually rendered (the plugin code). This is true for all keyboard interaction events.

This is already what we do in our code base in a lot of places and this is not the first HoC that suffered from this issue. I can cite WritingFlow component for instance.

Comment (ajlende) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: @youknowriad Thanks for the more specific example—it helped me know where to look. This is more complicated than it seems at surface level, so I started throwing together some more complete [examples on CodeSandbox](https://codesandbox.io/s/constrained-tabbing-lzzzp?file=/src/app.js) so I didn't just have to theorize about things. Then with Thanksgiving here in the US and a couple other things, I didn't get around to completing those until now. Sorry for the wait.

I came up with six scenarios, and I learned a few things when I was making them up. The full details are in the [README.md of the CodeSandbox](https://codesandbox.io/s/constrained-tabbing-lzzzp?file=/README.md).

tl;dr What I have (React events) actually works for every scenario<sup id="a1">[1](#f1)</sup>. Tabbable elements are queried with `@wordpress/dom` and the container catches the keypresses just fine either way, so event bubbling isn't actually a factor. Checking if the tabbables includes the event target is still necessary either way.

At this point, I'm mostly indifferent to which method is used. On one hand React events are simpler, "conventional wisdom" suggests that they should be preferred, and they work just fine. (Feel free to fork the CodeSandbox if you have an example which would break it.) But on the other hand, since the tabbable elements are queried from the DOM tree instead of the React tree, it might be more consistent to also rely on DOM events.

I'd appreciate it if you could look at the examples to make sure I didn't miss anything. Then, if you still prefer the DOM method, I'll copy over the code from the CodeSandbox to the PR. Thanks!

<sup><a id="f1">1</a></sup>There's some bugs described in the CodeSandbox README.md, but they're unrelated to `withConstrainedTabbing` and happen even when using DOM events. [⮌](#a1)

Comment (youknowriad) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: Hi There!  thanks for following-up. I forked your example to demonstrate the issues. See the fail.js file, you can try to switch the import of withConstrinedTabbing HoC. When you use the DOM, tab don't reach toggle 5 as expected but if you do use "virtual" tabbing is not contained anymore.

https://codesandbox.io/s/constrained-tabbing-forked-zqlsp?file=/src/fail.js

I believe in your original examples you forgot `bubblesVirtually` in your Slots (this prop is actually what triggers the portalling behavior, we have two different slot implementations)

Comment (youknowriad) [reply packages/components/src/higher-order/with-constrained-tabbing/index.js:37]: That said, since our last discussion we already refactored this HoC to a hook using useConstrainedTabbing (and it's dom based), so I believe this issue is probably solved.

**Comment by ajlende** [comment]
Closed by #27544
